File: algorithms/0019-remove-nth-node-from-end-of-list.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, x):
#         self.val = x
#         self.next = None

class Solution:
    def removeNthFromEnd(self, head: ListNode, n: int) -> ListNode:
        a = head
        for _ in range(n):
            a = a.next
        if a is None:
            return head.next
        b = head
        while a.next is not None:
            a, b = a.next, b.next
        b.next = b.next.next
        return head

File: algorithms/0019-remove-nth-node-from-end-of-list.go
/**
 * Definition for singly-linked list.
 * type ListNode struct {
 *     Val int
 *     Next *ListNode
 * }
 */
 func removeNthFromEnd(head *ListNode, n int) *ListNode {
	a := head
	for i := 0; i < n; i++ {
		a = a.Next
	}
	if a == nil {
		return head.Next
	}
	b := head
	for a != nil {
		a, b = a.Next, b.Next
	}
	b.Next = b.Next.Next
	return head
}
